lib/IRGen/IRGenModule.cpp

```diff
@@ -430,7 +430,24 @@ IRGenModule::IRGenModule(IRGenerator &irgen,
                     // point too.
   };
   ObjCBlockStructTy->setBody(objcBlockElts);
-  
+
+  // Class _Nullable callback(Class _Nonnull cls, void * _Nullable arg);
+  llvm::Type *params[] = { ObjCClassPtrTy, Int8PtrTy };
+  ObjCUpdateCallbackTy = llvm::FunctionType::get(ObjCClassPtrTy, params, false);
+
+  // The full class stub structure, including a word before the address point.
```

Comment:
Why do we need to do this? Is it because we treat it like normal Swift metadata, which has the address point at -1 for classes?

Reply:
It's to work around a linker issue. The linker gets confused when it sees categories that don't point to "real" Objective-C classes. @mikeash knows the details

Reply:
Short version: the linker tries to attach categories to classes at link time when possible. These are not real classes so that goes badly. This makes the class look like a Swift class (because of the data before the ObjC class pointer) which the linker doesn't try to optimize in this way.

Reply:
Is that going to be permanent ABI, or can we eliminate it in the future with a cooperative linker?

Reply:
We can eliminate it in the future. I think the linker will be able to start applying the category optimization to Swift classes now that Swift has settled down some (Apple folks see rdar://problem/48850004 for that). If it does this, then it will also need to identify these stubs by some other means, such as checking the faux-isa. At that point, we can ditch the pre-class padding word.